src/headerssync.cpp

```diff
@@ -0,0 +1,318 @@
+#include <headerssync.h>
+#include <logging.h>
+#include <util/check.h>
+#include <pow.h>
+#include <timedata.h>
+
+// Store a commitment to a header every HEADER_COMMITMENT_FREQUENCY blocks
+constexpr size_t HEADER_COMMITMENT_FREQUENCY = 571;
+// Pull out headers for acceptance during redownload once our buffer exceeds this amount
+constexpr size_t REDOWNLOAD_HEADERS_THRESHOLD = 25*HEADER_COMMITMENT_FREQUENCY;
+// Ensure that the redownload buffer is a multiple of the header commitment
+// frequency, so that we aren't wasting memory.
+static_assert(REDOWNLOAD_HEADERS_THRESHOLD % HEADER_COMMITMENT_FREQUENCY == 0);
+
+// Our memory analysis assumes 48 bytes for a CompressedHeader (so we should
+// re-calculate parameters if we compress further)
+static_assert(sizeof(CompressedHeader) == 48);
+
+HeadersSyncState::HeadersSyncState(NodeId id, const Consensus::Params& consensus_params) :
+    m_id(id), m_consensus_params(consensus_params)
+{ }
+
+// Free any memory in use, and mark this object as no longer usable.
+void HeadersSyncState::Finalize()
+{
+    Assume(m_download_state != State::FINAL);
+    m_header_commitments.clear();
+    m_last_header_received.SetNull();
+    m_blockhash_with_sufficient_work.SetNull();
+    std::deque<CompressedHeader>().swap(m_redownloaded_headers);
+    m_redownload_buffer_last_hash.SetNull();
+    m_redownload_buffer_first_prev_hash.SetNull();
+    m_process_all_remaining_headers = false;
+    m_current_height = 0;
+
+    m_download_state = State::FINAL;
+}
+
+// Initialize the parameters for this headers download, validate this first
+// batch, and request more headers.
+std::optional<CBlockLocator> HeadersSyncState::StartInitialDownload(const CBlockIndex* chain_start,
+        const std::vector<CBlockHeader>& initial_headers, const arith_uint256&
+        minimum_required_work, CBlockLocator&& chain_start_locator)
+{
+    // A new instance of this object should be instantiated for every headers
+    // sync, so that we don't reuse our salted hasher between syncs.
+    assert(m_download_state == State::INITIAL_DOWNLOAD);
```

Comment:
In commit "Utilize anti-DoS headers download strategy"

Should there be a restriction to not call StartInitialDownload twice on the same object?

Reply:
Seems reasonable, I added a variable to do that.  Let me know if that looks good now.

Reply:
Done.